### ISSUE TITLE
Configuration and database rework

### ISSUE THREAD

**Issue by eine** [filebrowser/filebrowser#489]
This issue summarizes the requested features for the workflow related to the configuration file, the database, how user profiles are set up, etc. Some of these are frequently shown in multiple issues and this is an effort to put them all together. It is also the place where we can add and track new related requests. 

There was some previous discussion about this in #371. However, rather than merging the configuration file and the database, I think that the configuration file should be extended ([#381issuecomment-373983968](https://github.com/filebrowser/filebrowser/issues/381#issuecomment-373983968)) to be used as a template to initialize/set the database (probably periodically).

## Config file

- [x] Define multiple users in the config file [#371issuecomment-373985636](https://github.com/filebrowser/filebrowser/issues/371#issuecomment-373985636). See go-yaml/yaml#394.
  - Use (additional) JSON file to configure users [#424issuecomment-392551151](https://github.com/filebrowser/filebrowser/issues/424#issuecomment-392551151)
- [x] allow to set commands for events in config file [#246issuecomment-349563766](https://github.com/filebrowser/filebrowser/issues/246#issuecomment-349563766)
- [x] Add option "default sort mode" [#326issue-287085355](https://github.com/filebrowser/filebrowser/issues/326#issue-287085355) (for each user)
- [x] Rules disallow/allow in config file or globally #424 (for each user)
- [x] More flexible rights/permission scheme #502
  - [x] Add user group features to set some permissions [#401issue-316476440](https://github.com/filebrowser/filebrowser/issues/401#issue-316476440). *We can try using YAML variables to define groups. But I don't know if these groups are readable*.
		- We support YAML. Please see #614 for actual user groups.
- [x] Customizing admin credentials via config #257. *Let `admin` be a group and not a user?*
- [ ] ~Support for password changing via config #381 (for each user)~ via CLI
- [ ] ~Allow disabling user submitted stylesheets #507 (per user)~ no per-user CSS anymore.

## CLI flags/options

- [x] Customizing admin credentials via flags #257, #381, #435.
- [ ] ~Add option to enabe/disable [watching-and-re-reading-config-files](https://github.com/spf13/viper#watching-and-re-reading-config-files).~
- [x] Subcommand/option to [#401issue-316476440](https://github.com/filebrowser/filebrowser/issues/401#issue-316476440):
- [x] Add users from the config file to the database without removing existing ones. If conflict:
  - Overwrite.
  - Show both in the terminal and ask which one to keep.
- [x] Remove one/some/all users from the database.
- [x] Dump all users from the database [#371issuecomment-370220675](https://github.com/filebrowser/filebrowser/issues/371#issuecomment-370220675), [#371issuecomment-374574825](https://github.com/filebrowser/filebrowser/issues/371#issuecomment-374574825).
- [x] Subcommand to encrypt a password? [#371issuecomment-370157182](https://github.com/filebrowser/filebrowser/issues/371#issuecomment-370157182)
- [x] Enable settings with `--no-auth` #349, #425. *Use username `no-auth` without passwd and provide settings in the config file as any other user?*
- [x] Initialize conf/db interactively [#371issuecomment-373989360](https://github.com/filebrowser/filebrowser/issues/371#issuecomment-373989360).

## Environment variables

- [x] Configuration using environment variables #360, [#257issuecomment-337812483](https://github.com/filebrowser/filebrowser/issues/257#issuecomment-337812483). *Implemented in #504 -> #506. See also [spf13/viper#working-with-environment-variables](https://github.com/spf13/viper#working-with-environment-variables)) and [precedence](https://github.com/spf13/viper#why-viper)*.
- [x] Admin credentials via envvars [#381issuecomment-380829350](https://github.com/filebrowser/filebrowser/issues/381#issuecomment-380829350), #435.

## API

- [x] Support for password changing [#381issuecomment-372309336](https://github.com/filebrowser/filebrowser/issues/381#issuecomment-372309336), #510.
- [ ] ~LDAP & Active Directory (AD) Sync #301.~ It's not a blocker for this issue and since we now have an `Auther` interface, anyone interested can contribute with it.

---

@hacdias, @Equim-chan, everyone, please let me know if your miss any feature/request that should belong here.

**Comment by vitaliy-kuzmich** [comment]
I'd like to notice, that you do not need database at all. Why not just keep it simple ? Because I tried nextcloud and had horrible experience. So all that is needed is just filebrowser, why not to use srandard unix/linux users and theirs home folders ?

**Comment by eine** [comment]
@vitaliy-kuzmich, filebrowser can be used in windows too, and docker images are based on scratch, so we cannot ensure that standard unix/linux users exists in the host.

**Comment by hacdias** [comment]
@1138-4EB I think we shall close this with 2.0

**Comment by eine** [comment]
Let's better review the items once v2 is ready for release. Most of them will be already achieved, but I am not sure about all of them.

**Comment by hacdias** [comment]
@1138-4EB I reviewed the items and checked the ones that are complete and removed some that I thought are not related or are invalid in the new version.

Things missing:

> LDAP & Active Directory (AD) Sync

We now have the support to add more kinds of Authentication. If someone wants to implement this, we're happy to receive a PR, but I don't believe it should be part of this issue, nor a blocker to launch v2.0.

> Add users from the config file to the database without removing existing ones. If conflict:

~Clean up `users import` to do what it says here.~ Done!

> Add user group features to set some permissions

I added support to yaml, but I believe that in the future we could actually introduce user groups. *But* for that I'm creating a new issue and moving it out of here since it's not a blocker for 2.0.

**Edit:** opened issue #614.

> Add option to enabe/disable watching-and-re-reading-config-files.

~Is this a good idea in the context of a server configuration?~ Moved to #617.
